kvdb/sqlite/db.go

```diff
@@ -0,0 +1,216 @@
+//go:build kvdb_sqlite && !(windows && (arm || 386)) && !(linux && (ppc64 || mips || mipsle || mips64))
+
+package sqlite
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+	"github.com/lightningnetwork/lnd/kvdb/sqlbase"
+	_ "modernc.org/sqlite" // Register relevant drivers.
+)
+
+const (
+	// sqliteOptionPrefix is the string prefix sqlite uses to set various
+	// options. This is used in the following format:
+	//   * sqliteOptionPrefix || option_name = option_value.
+	sqliteOptionPrefix = "_pragma"
+
+	// sqliteTxLockImmediate is dsn option used to ensure that write
+	// transactions are started immediately.
+	sqliteTxLockImmediate = "_txlock=immediate"
+)
+
+// NewSqliteBackend returns a db object initialized with the passed backend
+// config. If a sqlite connection cannot be established, then an error is
+// returned.
+func NewSqliteBackend(ctx context.Context, cfg *Config, dbPath, fileName,
+	prefix string) (walletdb.DB, error) {
+
+	// First, we add a set of mandatory pragma options to the query.
+	pragmaOptions := []struct {
+		name  string
+		value string
+	}{
+		{
+			name: "busy_timeout",
+			value: fmt.Sprintf(
+				"%d", cfg.BusyTimeout.Milliseconds(),
+			),
+		},
+		{
+			name:  "foreign_keys",
+			value: "on",
+		},
+		{
+			name:  "journal_mode",
+			value: "WAL",
+		},
+		{
+			name:  "synchronous",
+			value: "normal",
+		},
+		{
+			name:  "auto_vacuum",
+			value: "incremental",
+		},
+	}
+	sqliteOptions := make(url.Values)
+	for _, option := range pragmaOptions {
+		sqliteOptions.Add(
+			sqliteOptionPrefix,
+			fmt.Sprintf("%v=%v", option.name, option.value),
+		)
+	}
+
+	// Then we add any user specified pragma options.
+	for _, option := range cfg.PragmaOptions {
+		pair := strings.Split(option, "=")
+		if len(pair) != 2 {
+			return nil, fmt.Errorf("invalid pragma option: %v. "+
+				"Expected the format: name=value", option)
+		}
+
+		sqliteOptions.Add(
+			sqliteOptionPrefix,
+			fmt.Sprintf("%v=%v", pair[0], pair[1]),
+		)
+	}
+	fmt.Println("ELLE: ", filepath.Join(dbPath, fileName))
```

Comment:
Stray debug statement.

Reply:
oops 😬

Reply:
fixed - will try set up a "banned words" commit hook to avoid this in future